pkg/config/file_config.go

```diff
@@ -0,0 +1,143 @@
+	if rules.RuntimeConfigPath == "" {
+		rules.RuntimeConfigPath = runtimeConfigPathDefault
+	}
+	clusterConfig, err = v1beta1.ConfigFromFile(rules.RuntimeConfigPath, "")
+	if err != nil {
+		return nil, err
+	}
+	return clusterConfig, nil
+}
+
+// validate accepts a ClusterConfig object and validates config fields
+func (rules *ClientConfigLoadingRules) Validate(clusterConfig *v1beta1.ClusterConfig, k0sVars constant.CfgVars) (*v1beta1.ClusterConfig, error) {
+	if clusterConfig.Spec.Storage.Type == v1beta1.KineStorageType && clusterConfig.Spec.Storage.Kine == nil {
+		logrus.Warn("storage type is kine but no config given, setting up defaults")
+		clusterConfig.Spec.Storage.Kine = v1beta1.DefaultKineConfig(k0sVars.DataDir)
+	}
+	if clusterConfig.Spec.Install == nil {
+		clusterConfig.Spec.Install = v1beta1.DefaultInstallSpec()
+	}
+
+	errors := clusterConfig.Validate()
+	if len(errors) > 0 {
+		messages := make([]string, len(errors))
+		for _, e := range errors {
+			messages = append(messages, e.Error())
+		}
+		return nil, fmt.Errorf(strings.Join(messages, "\n"))
+	}
+	return clusterConfig, nil
+}
+
+// ParseRuntimeConfig parses the `--config` flag and generates a config object
+// it searches for the default config path. if it does not exist, and no other custom config-file is given, it will generate default config
+func (rules *ClientConfigLoadingRules) ParseRuntimeConfig() (*v1beta1.ClusterConfig, error) {
+	var cfg *v1beta1.ClusterConfig
+
+	if rules.RuntimeConfigPath == "" {
+		rules.RuntimeConfigPath = runtimeConfigPathDefault
+	}
+
+	// don't create the config file, if it already exists
+	if file.Exists(rules.RuntimeConfigPath) {
+		logrus.Infof("runtime config found: using %s", rules.RuntimeConfigPath)
+	}
+
+	switch CfgFile {
+	case "-":
+		// parse clusterConfig from stdin and write the contents to a temp directory
+		return rules.readFromStdin()
+	// no config-file is given, so either look for a config-file in the default location, or generate defaults
+	case constant.K0sConfigPathDefault:
+		// file doesn't exist, so we need to generate defaults
+		if rules.IsDefaultConfig() {
+			cfg = rules.generateDefaults()
+		} else {
+			return v1beta1.ConfigFromFile(constant.K0sConfigPathDefault, K0sVars.DataDir)
+		}
+	default:
+		return v1beta1.ConfigFromFile(CfgFile, K0sVars.DataDir)
+	}
+	return cfg, nil
+}
+
+// InitRuntimeConfig generates the runtime /run/k0s/k0s.yaml
+func (rules *ClientConfigLoadingRules) InitRuntimeConfig() error {
+	cfg, err := rules.ParseRuntimeConfig()
+	if err != nil {
+		return err
+	}
+
+	yamlData, err := yaml.Marshal(&cfg)
+	if err != nil {
+		return err
+	}
+
+	return rules.writeConfig(yamlData)
+}
+
+// generate default config and return the file path
+func (rules *ClientConfigLoadingRules) generateDefaults() (config *v1beta1.ClusterConfig) {
+	logrus.Info("no config file given, generating default config")
```

Comment:
maybe this could just be a `println` so you don't have to filter it out or fiddle with logrus output levels when you want to hide this from pipes.

or we could have separation for levels, warn...panic to stderr, trace..info to stdout, as documented here: https://github.com/sirupsen/logrus/tree/f104497f2b2129ab888fd274891f3a278756bcde/hooks/writer - then this could be `logrus.Warn`.